Library/Homebrew/utils/pypi.rb

```diff
@@ -8,49 +8,59 @@ module PyPI
   PYTHONHOSTED_URL_PREFIX = "https://files.pythonhosted.org/packages/"
   private_constant :PYTHONHOSTED_URL_PREFIX
 
-  # PyPI Package
-  #
+  # Represents a Python package.
+  # This package can be a PyPI package (either by name/version or PyPI distribution URL),
+  # or it can be a non-PyPI URL.
   # @api private
   class Package
-    attr_accessor :name, :extras, :version
-
     sig { params(package_string: String, is_url: T::Boolean).void }
     def initialize(package_string, is_url: false)
       @pypi_info = nil
+      @package_string = package_string
+      @is_url = is_url
+      @is_pypi_url = package_string.start_with? PYTHONHOSTED_URL_PREFIX
+    end
 
-      if is_url
-        match = if package_string.start_with?(PYTHONHOSTED_URL_PREFIX)
-          File.basename(package_string).match(/^(.+)-([a-z\d.]+?)(?:.tar.gz|.zip)$/)
-        end
-        raise ArgumentError, "Package should be a valid PyPI URL" if match.blank?
+    sig { returns(String) }
+    def name
+      @name ||= basic_metadata[0]
```

Comment:
What would you think about setting `@name` inside `basic_metadata` and calling out to it if unset? Feels a bit harder to follow what's being returned from it when it's separated up here.

Reply:
Sounds good to me; doing now!

Reply:
[0f40e22](https://github.com/Homebrew/brew/pull/15511/commits/0f40e224bdabf13520dfba333257e9a156b1661c) has these changes.